wallet-new/test/unit/UTxO/PreChain.hs

```diff
@@ -64,29 +73,82 @@ preChain = fmap (, ()) . DepIndep
 -- See 'fromPreChain'
 data FromPreChain h a = FromPreChain {
       -- | The boot transaction that was used in the translation
-      fpcBoot :: Transaction h Addr
+      fpcBoot   :: Transaction h Addr
 
       -- | The resulting chain (i.e., list of list of transactions)
       -- This does /not/ include the bootstrap transaction.
-    , fpcChain :: Chain h Addr
+    , fpcChain  :: Chain h Addr
 
       -- | The resulting ledger (i.e., flat list of transactions)
       -- This /does/ include the bootstrap transaction .
     , fpcLedger :: Ledger h Addr
 
       -- | Any additional information that was included in the 'PreChain'.
-    , fpcExtra :: a
+    , fpcExtra  :: a
     }
 
 fromPreChain :: (Hash h Addr, Monad m)
              => PreChain h m a -> TranslateT IntException m (FromPreChain h a)
 fromPreChain pc = do
     fpcBoot <- asks bootstrapTransaction
-    (txs, fpcExtra) <- calcFees fpcBoot =<< lift (runDepIndep pc fpcBoot)
+    (txs, fpcExtra) <- calcChainFees fpcBoot =<< lift (runDepIndep pc fpcBoot)
     let fpcChain  = Chain txs -- doesn't include the boot transactions
         fpcLedger = chainToLedger fpcBoot fpcChain
     return FromPreChain{..}
 
+{-------------------------------------------------------------------------------
+  Tree with some information still missing.
+-------------------------------------------------------------------------------}
+
+-- | Blocks organised in a tree structure.
+type BlockTree h a = OldestFirst Tree (UTxO.DSL.Block h a)
+
+-- | Generalisation of a 'PreChain'.
+--   A 'PreChain' is the flattening of a 'PreTree' with a single branch.
+type PreTree h m a = DepIndep (Transaction h Addr) (Tree [Fee]) m (BlockTree h Addr, a)
+
+preTree :: Functor m
+        => (Transaction h Addr -> m ((Tree [Fee]) -> BlockTree h Addr))
+        -> PreTree h m ()
+preTree = fmap (, ()) . DepIndep
+
+-- | Result of translating a 'PreTree'
+--
+-- See 'fromPreTree'
+-- Compared to 'FromPreChain':
+-- - we do not calculate a ledger, since we have not yet fixed on a single chain.
+--   Instead we include the list of all addresses in the tree. Note that this may
+--   include addresses which have no transactions to/from them in the final
+--   blockchain, because they were in blocks which were rolled back.
+-- - the boot transaction is included in the chain.
+data FromPreTree h a = FromPreTree {
+      -- | The resulting tree
+      fptTree      :: BlockTree h Addr
+
+      -- | The boot transaction
+    , fptBoot      :: Transaction h Addr
+
+      -- | Addresses involved in this block tree
+    , fptAddresses :: Set Addr
+
+      -- | Any additional information that was included in the 'PreChain'.
+    , fptExtra     :: a
```

Comment:
Making the fields strict, unless there's a specific reason not to, tends to be a good idea down the road as it save us from causing accidental space leaks.

Reply:
:+1:

Reply:
There's a minor pull in the other direction here also: we may want all this stuff to stream eventually. But how that will work isn't at all clear to me right now, so being strict is fine for now.